lib/blocks.php

```diff
@@ -216,3 +216,189 @@ function gutenberg_register_legacy_social_link_blocks() {
 	}
 }
 add_action( 'init', 'gutenberg_register_legacy_social_link_blocks' );
+
+/**
+ * Renders the classNames and styles for blocks
+ *
+ * @param string $block_content Output of the current block.
+ * @param array  $block Block Object.
+ * @return string New block output.
+ */
+function gutenberg_experimental_apply_classnames_and_styles( $block_content, $block ) {
+	if ( ! isset( $block['attrs'] ) ) {
+		return $block_content;
+	}
+
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	// If no render_callback, assume styles have been previously handled.
+	if ( ! $block_type || ! $block_type->render_callback ) {
+		return $block_content;
+	}
+	// Check what style features the block supports.
+	$supports = gutenberg_experimental_global_styles_get_supported_styles( $block_type->supports );
+
+	$attributes = array();
+	$attributes = gutenberg_experimental_build_css_colors( $attributes, $block['attrs'], $supports );
+	$attributes = gutenberg_experimental_build_css_typography( $attributes, $block['attrs'], $supports );
+
+	if ( ! count( $attributes ) ) {
+		return $block_content;
+	}
+
+	$dom = new DOMDocument( '1.0', 'utf-8' );
+
+	// Suppress warnings from this method from polluting the front-end.
+	// @codingStandardsIgnoreStart
+	if ( ! @$dom->loadHTML( $block_content, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD | LIBXML_COMPACT ) ) {
```

Comment:
Just for reference: this introduced #24445, with a likely fix in #24447. I don't find it surprising that we discover quirks in these PHP interfaces as we start doing more server-side manipulation of content. :)

Reply:
Thanks for catching it. 👍